cmd/promtool/main.go

```diff
@@ -641,3 +659,185 @@ func (j *jsonPrinter) printLabelValues(v model.LabelValues) {
+		if s == "" {
+			newSymbol = ""
+		}
+		newSymbolsMap[s] = newSymbol
+		allSymbols[newSymbol] = struct{}{}
+	}
+
+	indexw, err := index.NewWriter(outputFilePath)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return 1
+	}
+	defer indexw.Close()
+
+	// Write symbols.
+	if err := indexw.AddSymbols(allSymbols); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return 1
+	}
+
+	// Write Series.
+	posts, err := indexr.Postings(apkName, apkValue)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return 1
+	}
+
+	for posts.Next() {
+		p := posts.At()
+		labelsBuf = labelsBuf[:0]
+		chunksBuf = chunksBuf[:0]
+
+		if err := indexr.Series(p, &labelsBuf, &chunksBuf); err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return 1
+		}
+
+		// Recording the original labels values which is needed
+		// to fetch and write the postings.
+		for _, l := range labelsBuf {
+			valset, ok := values[l.Name]
+			if !ok {
+				valset = map[string]struct{}{}
+				values[l.Name] = valset
+			}
+			valset[l.Value] = struct{}{}
+		}
+
+		// Rewriting the labels.
+		for i := range labelsBuf {
+			labelsBuf[i].Name = newSymbolsMap[labelsBuf[i].Name]
+			labelsBuf[i].Value = newSymbolsMap[labelsBuf[i].Value]
+		}
+
+		if err := indexw.AddSeries(p, labelsBuf, chunksBuf...); err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return 1
+		}
+	}
+
+	// Write label indices and postings.
+	if err := writePostings(indexw, values, indexr, newSymbolsMap); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return 1
+	}
+
+	return 0
+}
+
+// writePostings writes the postings into the index file.
+func writePostings(indexw *index.Writer, values map[string]map[string]struct{}, indexr *index.Reader, symbolMap map[string]string) (err error) {
+	var (
+		maxNumValues      int
+		numLabelValues    int
+		names             = make([]string, 0, len(values)+1)
+		apkName, apkValue = index.AllPostingsKey()
+	)
+	for _, v := range values {
+		numLabelValues += len(v)
+		if len(v) > maxNumValues {
+			maxNumValues = len(v)
+		}
+	}
+
+	labelValuesBuf := make([]string, 0, maxNumValues)
+	for n, v := range values {
+		labelValuesBuf = labelValuesBuf[:0]
+		names = append(names, n)
+
+		if mappedName, ok := symbolMap[n]; ok {
+			n = mappedName
+		} else {
+			return fmt.Errorf("Mapped symbol not found for name: %s", n)
+		}
+
+		for val := range v {
+			if mappedVal, ok := symbolMap[val]; ok {
+				labelValuesBuf = append(labelValuesBuf, mappedVal)
+			} else {
+				return fmt.Errorf("Mapped symbol not found for value: %s", val)
+			}
+		}
+		if err := indexw.WriteLabelIndex([]string{n}, labelValuesBuf); err != nil {
+			return err
+		}
+	}
+	names = append(names, apkName)
+	values[apkName] = map[string]struct{}{apkValue: struct{}{}}
+	sort.Strings(names)
+
+	for _, n := range names {
+		labelValuesBuf = labelValuesBuf[:0]
+		for v := range values[n] {
+			labelValuesBuf = append(labelValuesBuf, v)
+		}
+		sort.Strings(labelValuesBuf)
+
+		if mappedName, ok := symbolMap[n]; ok {
```

Comment:
I think this lookup should be done after `posts, err := indexr.Postings(n, v)`. Otherwise, the `n` will be the new symbol and `posts, err := indexr.Postings(n, v)` will return an empty postings without any error (because the error will be nil when the name/value lookup fails).

Reply:
Yes, you are right. Thanks for pointing out!